In [23]:
import pandas as pd
import statsmodels.api as sm
import numpy as np

In [6]:
df = pd.read_excel("../../data/CandyData.xlsx")

In [8]:
df.head()

competitorname  chocolate  fruity  caramel  peanutyalmondy  nougat  \
0      100 Grand          1       0        1               0       0   
1   3 Musketeers          1       0        0               0       1   
2       One dime          0       0        0               0       0   
3    One quarter          0       0        0               0       0   
4      Air Heads          0       1        0               0       0   

   crispedricewafer  hard  bar  pluribus  sugarpercent  pricepercent  \
0                 1     0    1         0         0.732         0.860   
1                 0     0    1         0         0.604         0.511   
2                 0     0    0         0         0.011         0.116   
3                 0     0    0         0         0.011         0.511   
4                 0     0    0         0         0.906         0.511   

   winpercent  
0   66.971725  
1   67.602936  
2   32.261086  
3   46.116505  
4   52.341465

Перед выполнением задания создайте переменную winpercent_0_1, которая принимает значение 1, если рейтинг конфеты больше или равен 50, и 0 в обратном случае.

In [27]:
df["winpercent_0_1"] = np.where(df['winpercent'] >= 50, 1, 0)

Постройте логистическую модель, где зависимой переменной выступает winpercent_0_1, а независимой – sugarpercent. Выберите верную интерпретацию константы.

In [28]:
X = df["sugarpercent"]
y = df["winpercent_0_1"]
X_constant = sm.add_constant(X)
logit = sm.Logit(y, X_constant).fit()
print(logit.summary())

Optimization terminated successfully.
         Current function value: 0.672972
         Iterations 4
                           Logit Regression Results                           
Dep. Variable:         winpercent_0_1   No. Observations:                   85
Model:                          Logit   Df Residuals:                       83
Method:                           MLE   Df Model:                            1
Date:                Sun, 27 Mar 2022   Pseudo R-squ.:                 0.02433
Time:                        20:29:05   Log-Likelihood:                -57.203
converged:                       True   LL-Null:                       -58.629
Covariance Type:            nonrobust   LLR p-value:                   0.09122
                   coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
const           -0.8058      0.447     -1.803      0.071      -1.682       0.070
sugarpercent     1.3276

C:\Users\dvavk\anaconda3\lib\site-packages\statsmodels\tsa\tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


In [31]:
print(np.exp(-0.8058))

0.4467303992481016


Если предиктор = 0, т.е, что у конфеты будет рейтинг = 0 (низкий), то шансы = 0.45

Постройте логистическую модель, где зависимой переменной выступает winpercent_0_1, а независимой – sugarpercent. Выберите верную интерпретацию коэффициента при объясняющей переменной. Выберите один верный ответ.

In [32]:
print(np.exp(1.3276))

3.771979764137272


т.е при увеличении процента сахара на 1 шанс, что у конфеты будет рейтинг = 1 (высокий) = 3.77

Укажите, верно ли следующее утверждение: на уровне значимости 0.05 процент сахара в конфете статистически значимо влияет на рейтинг конфеты.

Неверно, т.к.  

                P>|z|

const           0.071  
sugarpercent    0.096  

Чему будет равен псевдо R-квадрат в данной модели? Выберите один верный ответ.

Pseudo R-squ.:                 0.02433

Постройте логистическую модель, где зависимой переменной выступает winpercent_0_1, а независимыми – sugarpercent и pricepercent. Чему будет равено отношение шансов при цене конфеты? Введите ответ в виде десятичной дроби, округлите до двух знаков после запятой.

In [43]:
X = pd.concat([df["sugarpercent"], df["pricepercent"]], axis=1)
X.head()

sugarpercent  pricepercent
0         0.732         0.860
1         0.604         0.511
2         0.011         0.116
3         0.011         0.511
4         0.906         0.511

In [44]:
y = df["winpercent_0_1"]
X_constant = sm.add_constant(X)
logit = sm.Logit(y, X_constant).fit()
print(logit.summary())

Optimization terminated successfully.
         Current function value: 0.619123
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:         winpercent_0_1   No. Observations:                   85
Model:                          Logit   Df Residuals:                       82
Method:                           MLE   Df Model:                            2
Date:                Sun, 27 Mar 2022   Pseudo R-squ.:                  0.1024
Time:                        21:11:28   Log-Likelihood:                -52.625
converged:                       True   LL-Null:                       -58.629
Covariance Type:            nonrobust   LLR p-value:                  0.002470
                   coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
const           -1.6935      0.580     -2.918      0.004      -2.831      -0.556
sugarpercent     0.5896

C:\Users\dvavk\anaconda3\lib\site-packages\statsmodels\tsa\tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


In [45]:
print(np.exp(2.6012))

13.479904218413797


Чему будет равно отношение шансов при проценте сахара в конфете? Введите ответ в виде десятичной дроби, округлите до двух знаков после запятой.

In [46]:
print(np.exp(0.5896))

1.8032669643315304


Укажите, верно ли следующее утверждение: на уровне значимости 0.05 цена конфеты статистически значимо влияет на ее рейтинг.

Верно, т.к.  

                P>|z|

pricepercent    0.004

Во сколько раз будет ниже шанс высокого рейтинга конфеты при нулевом содержании сахара и нулевом значении цены? Введите ответ в виде десятичной дроби, округлите до двух знаков после запятой.

In [47]:
print(np.exp(-1.6935))

0.18387483452370543


Постройте логистическую модель, где зависимой переменной выступает winpercent_0_1, а независимыми – центрированные sugarpercent и pricepercent. Во сколько раз будет ниже шанс высокого рейтинга конфеты при среднем содержании сахара и среднем значении цены? Введите ответ в виде десятичной дроби, округлите до двух знаков после запятой.

In [48]:
X = pd.concat([df["sugarpercent"]-df["sugarpercent"].mean(), df["pricepercent"]-df["pricepercent"].mean()], axis=1)
y = df["winpercent_0_1"]
X_constant = sm.add_constant(X)
logit = sm.Logit(y, X_constant).fit()
print(logit.summary())

Optimization terminated successfully.
         Current function value: 0.619123
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:         winpercent_0_1   No. Observations:                   85
Model:                          Logit   Df Residuals:                       82
Method:                           MLE   Df Model:                            2
Date:                Sun, 27 Mar 2022   Pseudo R-squ.:                  0.1024
Time:                        21:23:00   Log-Likelihood:                -52.625
converged:                       True   LL-Null:                       -58.629
Covariance Type:            nonrobust   LLR p-value:                  0.002470
                   coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
const           -0.1916      0.235     -0.816      0.414      -0.652       0.269
sugarpercent     0.5896

C:\Users\dvavk\anaconda3\lib\site-packages\statsmodels\tsa\tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


In [49]:
print(np.exp(-0.1916))

0.8256370572724328


Посмотрите на результаты двух моделей (с одной центрированной объясняющей переменной sugarpercent и с двумя центрированными объясняющими переменными sugarpercent и pricepercent. Укажите, верно ли следующее утверждение: у модели с большим количеством объясняющих переменных качество выше. Обратите внимание на на псевдо R-квадрат.

In [50]:
X = pd.concat([df["sugarpercent"]-df["sugarpercent"].mean()], axis=1)
y = df["winpercent_0_1"]
X_constant = sm.add_constant(X)
logit = sm.Logit(y, X_constant).fit()
print(logit.summary())

Optimization terminated successfully.
         Current function value: 0.672972
         Iterations 4
                           Logit Regression Results                           
Dep. Variable:         winpercent_0_1   No. Observations:                   85
Model:                          Logit   Df Residuals:                       83
Method:                           MLE   Df Model:                            1
Date:                Sun, 27 Mar 2022   Pseudo R-squ.:                 0.02433
Time:                        21:26:04   Log-Likelihood:                -57.203
converged:                       True   LL-Null:                       -58.629
Covariance Type:            nonrobust   LLR p-value:                   0.09122
                   coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
const           -0.1704      0.221     -0.769      0.442      -0.604       0.264
sugarpercent     1.3276

C:\Users\dvavk\anaconda3\lib\site-packages\statsmodels\tsa\tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


у модели с двумя переменными  
Pseudo R-squ.: 0.1024  
у модели с одной переменной  
Pseudo R-squ.: 0.02433  

Верно, т.к. 0.1024 > 0.02433